File: frontend/src/types/portfolio.ts
// Types shared between the client (src/) and the server (server/).

export interface SocialLinks {
  github?: string;
  linkedin?: string;
  twitter?: string;
  website?: string;
}

export interface Stat {
  label: string;
  value: string;
}

export interface SeoMeta {
  /** <meta name="description"> and og:description. */
  metaDescription: string;
  keywords: string[];
  /** Absolute URL of a social preview image (og:image). Optional. */
  ogImage?: string;
}

export interface Profile {
  name: string;
  title: string;
  tagline: string;
  bio: string;
  location: string;
  email: string;
  phone?: string;
  /** Optional portrait URL; the hero shows a monogram blob when unset. */
  photoUrl?: string;
  links: SocialLinks;
  stats: Stat[];
  seo: SeoMeta;
}

/**
 * A pitch targeted at one visitor type (recruiter, freelance client, fellow
 * engineer, …). Drives the hero headline, the "Why Me" section, and the
 * primary call to action.
 */
export interface AudiencePitch {
  id: string;
  /** Switcher button text, e.g. "HIRING FOR A TEAM". */
  label: string;
  headline: string;
  pitch: string;
  /** "Title — detail" per entry; title is emphasised in the UI. */
  valueProps: string[];
  ctaLabel: string;
  ctaHref: string;
}

export interface Testimonial {
  id: string;
  quote: string;
  author: string;
  role: string;
  /** 1-5 stars; defaults to 5 in the UI when unset. */
  rating?: number;
}

export interface Service {
  id: string;
  title: string;
  /** Emoji shown in the service card icon circle. */
  emoji: string;
  /** One-liner for cards and meta descriptions. */
  summary: string;
  /** Long-form copy for the service details page; blank line = new paragraph. */
  description: string;
  deliverables: string[];
  tech: string[];
}

export interface PricingPlan {
  id: string;
  name: string;
  price: string;
  /** e.g. "/ hour", "/ project", "/ month" */
  unit: string;
  features: string[];
  highlighted: boolean;
}

export interface BlogPost {
  id: string;
  /** URL path segment: /blog/<slug> */
  slug: string;
  title: string;
  excerpt: string;
  /**
   * Lightweight markup: blank line = paragraph break, lines starting with
   * "## " = heading, lines starting with "- " = bullet list item.
   */
  content: string;
  date: string; // ISO date
  tags: string[];
  coverUrl?: string;
}

/**
 * A side of life beyond client work — content creation, founding startups,
 * SaaS products. Shown as "Beyond the code" cards; coming-soon until live.
 */
export interface Venture {
  id: string;
  title: string;
  emoji: string;
  tagline: string;
  description: string;
  live: boolean;
  /** Destination when live (channel, product site, …). */
  url?: string;
}

export interface ContactMessage {
  id: string;
  name: string;
  email: string;
  phone?: string;
  interest?: string;
  budget?: string;
  message: string;
  sentAt: string; // ISO timestamp
  read: boolean;
  status: 'new' | 'read' | 'replied' | 'converted' | 'archived';
  clientId?: string;
}

export interface Skill {
  id: string;
  name: string;
  category: string;
  level: number; // 0-100
  /** Emoji shown in the tools grid tile. */
  emoji?: string;
}

export interface Experience {
  id: string;
  company: string;
  role: string;
  period: string;
  location: string;
  highlights: string[];
  tech: string[];
}

export interface Project {
  id: string;
  title: string;
  tag: string;
  description: string;
  /** Long-form copy for the project details page; blank line = new paragraph. */
  details?: string;
  tech: string[];
  liveUrl?: string;
  repoUrl?: string;
}

export interface Education {
  id: string;
  institution: string;
  degree: string;
  period: string;
  detail?: string;
}

export interface Achievement {
  id: string;
  title: string;
  description: string;
  year: string;
}

export interface FaqEntry {
  id: string;
  question: string;
  answer: string;
}

export interface PortfolioData {
  profile: Profile;
  audiences: AudiencePitch[];
  services: Service[];
  pricing: PricingPlan[];
  testimonials: Testimonial[];
  skills: Skill[];
  experiences: Experience[];
  projects: Project[];
  blogPosts: BlogPost[];
  ventures: Venture[];
  education: Education[];
  achievements: Achievement[];
  faqs: FaqEntry[];
}

export type SectionKey = keyof PortfolioData;

export const SECTION_KEYS: SectionKey[] = [
  'profile',
  'audiences',
  'services',
  'pricing',
  'testimonials',
  'skills',
  'experiences',
  'projects',
  'blogPosts',
  'ventures',
  'education',
  'achievements',
  'faqs',
];

export interface UnansweredQuestion {
  id: string;
  question: string;
  askedAt: string; // ISO timestamp
  status: 'open' | 'answered' | 'dismissed';
  answer?: string;
}

// --- API payloads ---

export interface ChatRequest {
  message: string;
}

export interface ChatResponse {
  reply: string;
  /** Where the answer came from. */
  source: 'faq' | 'ai' | 'unanswered';
}

export interface LoginRequest {
  password: string;
}

export interface LoginResponse {
  token: string;
}

export interface UpdateSectionRequest<K extends SectionKey = SectionKey> {
  section: K;
  value: PortfolioData[K];
}

export interface AnswerQuestionRequest {
  answer: string;
  addToFaq: boolean;
}

File: backend/app/schemas/portfolio.py
from __future__ import annotations

from pydantic import BaseModel


class SocialLinks(BaseModel):
    github: str | None = None
    linkedin: str | None = None
    twitter: str | None = None
    website: str | None = None


class Stat(BaseModel):
    label: str
    value: str


class SeoMeta(BaseModel):
    metaDescription: str = ""
    keywords: list[str] = []
    ogImage: str | None = None


class Profile(BaseModel):
    name: str = ""
    title: str = ""
    tagline: str = ""
    bio: str = ""
    location: str = ""
    email: str = ""
    phone: str | None = None
    photoUrl: str | None = None
    links: SocialLinks = SocialLinks()
    stats: list[Stat] = []
    seo: SeoMeta = SeoMeta()


class AudiencePitch(BaseModel):
    id: str
    label: str
    headline: str
    pitch: str
    valueProps: list[str] = []
    ctaLabel: str = ""
    ctaHref: str = ""


class Testimonial(BaseModel):
    id: str
    quote: str
    author: str
    role: str = ""
    rating: int | None = None


class Service(BaseModel):
    id: str
    title: str
    emoji: str = ""
    summary: str = ""
    description: str = ""
    deliverables: list[str] = []
    tech: list[str] = []


class PricingPlan(BaseModel):
    id: str
    name: str
    price: str = ""
    unit: str = ""
    features: list[str] = []
    highlighted: bool = False


class BlogPost(BaseModel):
    id: str
    slug: str
    title: str
    excerpt: str = ""
    content: str = ""
    date: str = ""
    tags: list[str] = []
    coverUrl: str | None = None


class Venture(BaseModel):
    id: str
    title: str
    emoji: str = ""
    tagline: str = ""
    description: str = ""
    live: bool = False
    url: str | None = None


class ContactMessage(BaseModel):
    id: str
    name: str
    email: str
    phone: str | None = None
    interest: str | None = None
    budget: str | None = None
    message: str
    sentAt: str
    read: bool


class Skill(BaseModel):
    id: str
    name: str
    category: str = ""
    level: int = 0
    emoji: str | None = None


class Experience(BaseModel):
    id: str
    company: str
    role: str
    period: str = ""
    location: str = ""
    highlights: list[str] = []
    tech: list[str] = []


class Project(BaseModel):
    id: str
    title: str
    tag: str = ""
    description: str = ""
    details: str | None = None
    tech: list[str] = []
    liveUrl: str | None = None
    repoUrl: str | None = None
    coverUrl: str | None = None


class Education(BaseModel):
    id: str
    institution: str
    degree: str
    period: str = ""
    detail: str | None = None


class Achievement(BaseModel):
    id: str
    title: str
    description: str = ""
    year: str = ""


class FaqEntry(BaseModel):
    id: str
    question: str
    answer: str


class PortfolioData(BaseModel):
    profile: Profile
    audiences: list[AudiencePitch] = []
    services: list[Service] = []
    pricing: list[PricingPlan] = []
    testimonials: list[Testimonial] = []
    skills: list[Skill] = []
    experiences: list[Experience] = []
    projects: list[Project] = []
    blogPosts: list[BlogPost] = []
    ventures: list[Venture] = []
    education: list[Education] = []
    achievements: list[Achievement] = []
    faqs: list[FaqEntry] = []


class UnansweredQuestion(BaseModel):
    id: str
    question: str
    askedAt: str
    status: str = "open"
    answer: str | None = None


SECTION_KEYS = [
    "profile", "audiences", "services", "pricing", "testimonials",
    "skills", "experiences", "projects", "blogPosts", "ventures",
    "education", "achievements", "faqs",
]


class SectionUpdateRequest(BaseModel):
    section: str
    value: dict | list


class AnswerQuestionRequest(BaseModel):
    answer: str
    addToFaq: bool = False
